discovery/gossiper.go

```diff
@@ -323,6 +323,112 @@ func (d *AuthenticatedGossiper) ProcessLocalAnnouncement(msg lnwire.Message,
 	return nMsg.err
 }
 
+// ChannelUpdateID is a unique identifier for ChannelUpdate messages, as
+// channel updates can be identified by the (ShortChannelID, Flags)
+// tuple.
+type ChannelUpdateID struct {
+	// channelID represents the set of data which is needed to
+	// retrieve all necessary data to validate the channel existence.
+	channelID lnwire.ShortChannelID
+
+	// Flags least-significant bit must be set to 0 if the creating node
+	// corresponds to the first node in the previously sent channel
+	// announcement and 1 otherwise.
+	Flags uint16
+}
+
+// deDupedAnnouncements de-duplicates announcements that have been
+// added to the batch. Internally, announcements are stored in three maps
+// (one each for channel announcements, channel updates, and node
+// announcements). These maps keep track of unique announcements and
+// ensure no announcements are duplicated.
+type deDupedAnnouncements struct {
+	// channelAnnouncements are identified by the short channel id field.
+	channelAnnouncements map[lnwire.ShortChannelID]lnwire.Message
+
+	// channelUpdates are identified by the channel update id field.
+	channelUpdates map[ChannelUpdateID]lnwire.Message
+
+	// nodeAnnouncements are identified by node id field.
+	nodeAnnouncements map[*btcec.PublicKey]lnwire.Message
```

Comment:
It may not be immediately obvious, but this wouldn't properly de-duplicate the `NodeAnnouncement` for a particular public key if the `NodeAnnouncement` repeatedly added is a unique allocation each time. Using pointers as a key to a map is something that must be done carefully, otherwise the resulting behavior may not be what is expected. 

Instead the key to the map should be a `[33]byte`, or just the `vertex` type which is already present within this package. A `vertex` can be created with a `newVertex` function taking a `*btcec.PublicKey` as its sole argument. The tests should also be updated in order to assert that when adding multiple node announcements each from distinct allocations (with the same advertising public key), then assert that each instance was properly de-duplicated